### PR TITLE
Allow Multiple Invites to Party

### LINKED ISSUE #7792: Allow Multiple Invites
Right now, there is a problem where if you try to send someone an invite but they have another pending, you cannot send the invitation. Users should be allowed to have multiple invites pending, and when they choose a Party, the other Party invites should be automatically rejected.  (Note that this should not happen with Guilds.)

### PR DESCRIPTION
**Pull request by Yutsuten**
[//]: # (Note: See http://habitica.wikia.com/wiki/Using_Habitica_Git#Pull_Request for more info)

[//]: # (Put Issue # or URL here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes #7792

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)
This will change the number of possible party invitations a user can receive from one to unlimited.

- [x] (Server) Add array for invitations
- [x] (Server) Update references: from ``invitations.party`` to ``invitations.parties``
- [x] (Client) Show all invitations to join parties
- [x] Create tests

### Screenshot

Partial Result

[//]: # (Put User ID in here - found in Settings -> API)

----
UUID: 203f208a-e0f0-4442-bf16-d4f19e553e8a

### CODE REVIEW

File: website/server/libs/clearPartyInvitation.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+    if (party.id === groupID) {
+      partiesArr.splice(index, 1);
+    }
+  });
+
+  const lastInviteIndex = user.invitations.parties.length - 1;
+  user.invitations.party = user.invitations.parties[lastInviteIndex] || {};
+
+  user.markModified('invitations.party');
+  user.markModified('invitations.parties');
```
Comment (paglias) [inline website/server/libs/clearPartyInvitation.js]: is this necessary? Changes should be tracked automatically unless you modify a property of a Mixed field

File: website/server/libs/clearPartyInvitation.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
+// Transitioning invitation storage to an array of objects (see issue #7792 for reference)
+module.exports = function clearPartyInvitation (user, groupID) {
```
Comment (paglias) [inline website/server/libs/clearPartyInvitation.js]: export default function...?

**Comment by Yutsuten** [comment]
I haven't changed the visual, but if someone have great ideas in mind, please tell me so I change it!
That's how it is looking right now:

Partial Result

The ``invitations.party`` is still being updated with the lastest invitation, so I believe it won't break other applications that use the habitica API.
The ``invitations.parties`` is the brand new array of party invitations.

Now I only need to create some tests for this...
How about the migration? I'll do it or someone else? If it's me, can someone teach me how I create one?

File: website/common/locales/en/groups.json [hunk 3]
```diff
@@ -35,6 +35,8 @@
   "leave": "Leave",
   "invitedTo": "Invited to <%= name %>",
   "invitedToNewParty": "You were invited to join a party! Do you want to leave this party and join <%= partyName %>?",
+  "partyInvitationsText": "You have <%= numberInvites %> party invitations! Choose wisely, because you can only be in one party at a time.",
+  "joinPartyConfirmationText": "Are you sure you want to join the party \"<%= partyName %>\"? You can only be in one party at a time. If you join, all other party invitations will be cancelled.",
```
Comment (khipkin) [inline website/common/locales/en/groups.json]: Suggestion/nit: "canceled" should be changed to "rejected" to match the button text for rejecting an invitation.

File: website/server/controllers/api-v3/groups.js [hunk 4]
```diff
@@ -471,29 +471,34 @@ api.joinGroup = {
...
+        }
 
-      // If user was in a different party (when partying solo you can be invited to a new party)
-      // make him leave that party before doing anything
-      if (user.party._id) {
-        let userPreviousParty = await Group.getGroup({user, groupId: user.party._id});
-        if (userPreviousParty) await userPreviousParty.leave(user);
-      }
+        // If user was in a different party (when partying solo you can be invited to a new party)
+        // make him leave that party before doing anything
```
Comment (khipkin) [inline website/server/controllers/api-v3/groups.js]: "make **them** leave"

File: website/common/locales/en/groups.json (hunk 3 above)
Comment (Alys) [reply website/common/locales/en/groups.json]: Agreed. That makes it clear that the action taken automatically is the same as can be done manually.

File: website/server/controllers/api-v3/groups.js (hunk 4 above)
Comment (Alys) [reply website/server/controllers/api-v3/groups.js]: Ah yes, thanks @khipkin.  @Yutsuten We use [they / them as a singular pronoun](https://en.wikipedia.org/wiki/Singular_they) throughout Habitica to cover people of all genders.

Comment (Yutsuten) [reply website/server/controllers/api-v3/groups.js]: Oh I haven't even noticed about that. Thank you, I'll change.

File: website/common/locales/en/groups.json (hunk 3 above)
Comment (Yutsuten) [reply website/common/locales/en/groups.json]: Okay!

File: website/server/models/group.js [hunk 5]
```diff
@@ -376,6 +376,7 @@ schema.methods.removeGroupInvitations = async function removeGroupInvitations ()
   let userUpdates = usersToRemoveInvitationsFrom.map(user => {
     if (group.type === 'party') {
       user.invitations.party = {};
+      user.invitations.parties = [];
```
Comment (paglias) [inline website/server/models/group.js]: this shouldn't remove only one invitation (matching the group variable) instead of all?

**Review by paglias — CHANGES_REQUESTED** [review]

File: website/server/models/group.js (hunk 5 above)
Comment (Yutsuten) [reply website/server/models/group.js]: Yes, I'll fix it... And I noticed that there are still a lot of references to ``invitations.party`` in the code, so I need to update them all...

**Comment by Yutsuten** [comment]
Sorry for not being so active lately, I've got very busy...

Anyway, I think it is done. Tested multiple invites both when the user is not in a party and when he/she already is in a party (playing solo).

So, I'm ready for reviews and tips.

**Comment by Yutsuten** [comment]
Any reviews?

File: website/server/models/group.js [hunk 6]
```diff
@@ -376,6 +376,7 @@ schema.methods.removeGroupInvitations = async function removeGroupInvitations ()
   let userUpdates = usersToRemoveInvitationsFrom.map(user => {
     if (group.type === 'party') {
       user.invitations.party = {};
```
Comment (paglias) [inline website/server/models/group.js]: this should be filled with any remaining invitation to other parties?

**Review by paglias — DISMISSED** [review]

File: website/server/models/group.js (hunk 6 above)
Comment (Yutsuten) [reply website/server/models/group.js]: Oh, yes. I'll change it.

**Comment by Yutsuten** [comment]
Bump

**Comment by TheHollidayInn** [comment]
It seems your tests are failing. We can review after those are fixed. I can't look at what is failing right now.

**Comment by Yutsuten** [comment]
The error is:
```/home/travis/build/HabitRPG/habitica/website/client/components/inventory/item.vue```
 ```40:19  error  Missing trailing comma  comma-dangle```

But I haven't touched item.vue...

**Comment by paglias** [comment]
@Yutsuten it may be an issue that already existed when you forked the develop branch. Can you try to update develop and merge it into your branch?

**Comment by Yutsuten** [comment]
Strange... Other commits passed the tests...
Anyway, I updated my branch with develop.

**Comment by Yutsuten** [comment]
Bump

**Comment by Yutsuten** [comment]
Bump

**Review by SabreCat — APPROVED** [review]

**Comment by Alys** [comment]
Did this end up not being a breaking change for the API? I.e., will all third-party tools that use the API with its current code still continue to work in exactly the same way after this is released?

If there's any changes to the API's interface or results, we'd want to announce this well in advance of it being merged.

I'm sorry, I know I should be able to work this out myself from looking at the code but I'm really short on time these days. :(

**Comment by SabreCat** [comment]
@Alys Based on my read, there is one substantial difference: if you're checking for a party invite, you need to look through the `invitations.parties` array instead of the singular `invitations.party`. The server will handle the usual "send invitation" and "join group" actions without any changes to API calls, but if you fetch the user to check for invites, expectations will change there. We can discuss this in developer chat!

**Comment by SabreCat** [comment]
Upon a further look, checking `invitations.party` will still work, as that will hold the most recent invite received: https://github.com/HabitRPG/habitica/pull/8683/files#diff-b950a7c557b35ea58bc2685532c46170R916

**Comment by SabreCat** [comment]
One concern while testing: if you're already in a party, you get double confirm prompts--might be a bit of an annoyance? 

Also, it's kinda hard to tell the difference between invites if you're already in a party:

image

These might not be a big deal pending the frontend redesign, though.

**Comment by SabreCat** [comment]
I'm OK with this as is, but will leave it Merge Pending for a bit in case other devs want to weigh in on the concerns above.

**Comment by SabreCat** [comment]
Thanks for your effort and your patience with the review process, @Yutsuten! Welcome to Blacksmith tier 4!

**Comment by Yutsuten** [comment]
Oh great!!

**Pull request merged**